src/mds/Server.cc

```diff
@@ -1223,9 +1236,23 @@ void Server::find_idle_sessions()
 	break;
       }
       to_evict.push_back(session);
+      if (defer_client_eviction) {
+	laggy_clients.insert(session->get_client());
+      }
     }
   }
 
+  // don't evict client(s) if osds are laggy
```

Comment:
> How is this check guaranteeing that the OSDs are laggy? Shouldn't this check for `defer_client_eviction`?

`laggy_clients` will have entries only if `defer_client_eviction` is set, `defer_client_eviction` will be set only if the config option is set && any osd is laggy.https://github.com/ceph/ceph/pull/49971/files#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R1137-R1140

Reply:
> And I still think `to_evict` can be done away with. Can you explain why it isn't a goo approach (as per [#49971 (comment)](https://github.com/ceph/ceph/pull/49971#discussion_r1189655675))?

laggy_clients is a special scenario, will only come into action if config is set + any osd is laggy else the normal eviction will prevail using to_evict.

if to_evict is removed, in case defer_client_eviction is false (either config in unset or no laggy osd or both), there will be no client eviction (as laggy_clients is touched only if defer_client_eviction is true), which is not expected.

Reply:
removing to_evict will require modifying logic for client eviction in find_idle_sessions() https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1231-L1249 because to_evict stores sessions while laggy_clients stores clients and I want to do minimal changes to the existing code.

also it easy to read, both the cases can easily be bifurcated, i.e. clients in case where eviction must prevail and clients in case eviction must be deferred because some osd is laggy

Reply:
Discussed on IRC regarding tracking client_t vs Session* - the way other other interface are written, its fine to track client_t additionally.

Reply:
done